Blog post on Elastic's donation of the eBPF profiling agent (#4613)
open-telemetry/opentelemetry.io: content/en/blog/2024/elastic-contributes-continuous-profiling-agent.md

---
title: Elastic Contributes its Continuous Profiling Agent to OpenTelemetry
linkTitle: Elastic Contributes Profiling Agent # Mandatory, make sure that your short title.

> date: 2024-06-06 # Put the current date, we will keep the date updated until your PR is merged

Reviewer: ```suggestion
date: 2024-06-07
```

[based on eBPF](https://ebpf.io/) and by that a whole system, always-on solution
that observes code and third-party libraries, kernel operations, and other code
you don't own. It eliminates the need for code instrumentation
(run-time/bytecode), recompilation, or service restarts with low overhead, low
CPU (~1%), and memory usage in production environments.

The donated profiling agent facilitates identifying non-optimal code paths,
uncovering "unknown unknowns", and provides comprehensive visibility into the
runtime behavior of all applications. The continuous profiling agent provides
support for a wide range of runtimes and languages, such as:

- C/C++
- Rust
- Zig
- Go
- Java
- Python
- Ruby
- PHP
- Node.js / V8
- Perl
- .NET

## Benefits to OpenTelemetry

This contribution not only boosts the standardization of continuous profiling
for observability but also accelerates its adoption as a key signal in
OpenTelemetry. Customers benefit from a vendor-agnostic method of collecting
profiling data correlating it with existing signals, like tracing, metrics, and
logs, opening new potential for observability insights and a more efficient
troubleshooting experience.

> ### User Benefits of OpenTelemetry Profiling

Reviewer: ```suggestion
### User benefits of OpenTelemetry Profiling
```

---
title: Elastic Contributes its Continuous Profiling Agent to OpenTelemetry
linkTitle: Elastic Contributes Profiling Agent # Mandatory, make sure that your short title.
date: 2024-06-06 # Put the current date, we will keep the date updated until your PR is merged

> author:

Reviewer: ```suggestion
# prettier-ignore
cSpell:ignore: Bahubali Shetti Dmitry Filimonov Geisendörfer Halliday Christos Kalkanis
author:
```

  >- # If you have only one author, then add the single name on this line in quotes.
  [Bahubali Shetti](https://github.com/bshetti) (Elastic), [Alexander
  Wert](https://github.com/AlexanderWert) (Elastic), [Morgan
  McLean](https://github.com/mtwo) (Splunk), [Ryan
  Perry](https://github.com/Rperry2174) (Grafana)
draft: true # TODO: remove this line once your post is ready to be published
# canonical_url: http://somewhere.else/ # TODO: if this blog post has been posted somewhere else already, uncomment & provide the canonical URL here.
issue: https://github.com/open-telemetry/community/issues/1918
sig: Profiling SIG
---

Following significant collaboration between
[Elastic](https://www.elastic.co/observability-labs/blog/elastic-donation-proposal-to-contribute-profiling-agent-to-opentelemetry)

> and [OpenTelemetry's profiling community](./profiling.md), which included a

Reviewer: ```suggestion
and [OpenTelemetry's profiling community](/blog/2024/profiling/), which included a
```

I asssume you want to link to the profiling blog post?

Author: yes

thorough review process, we’re excited to announce that the OpenTelemetry
project has accepted
[Elastic's donation of its continuous profiling agent](https://github.com/open-telemetry/community/issues/1918).

This marks a significant milestone in establishing profiling as a core telemetry

> signal in OpenTelemetry. Elastic’s [eBPF](https://ebpf.io/) profiling agent

Reviewer: ```suggestion
signal in OpenTelemetry. Elastic’s [eBPF based](https://ebpf.io/) profiling agent
```

to vendor neutrality and community-driven development enhancing the overall
profiling and observability ecosystems.

The Elastic team would like to thank the OpenTelemetry community for the great
and constructive cooperation on the donation proposal. We look forward to
jointly establishing continuous profiling as an integral part of OpenTelemetry.

With today’s acceptance, Elastic’s continuous profiling agent will be
contributed to OpenTelemetry. This agent will now be jointly supported by both
Elastic’s team as well as a diverse set of official maintainers from different
companies:

- Dmitry Filimonov (Grafana Labs)
- Felix Geisendörfer (Datadog)
- Jonathan Halliday (Red Hat)
- Christos Kalkanis (Elastic)

## What is continuous profiling?

[Continuous profiling](https://www.cncf.io/blog/2022/05/31/what-is-continuous-profiling/)
is a technique used to understand the behavior of a software application by
collecting information about its execution. This includes tracking the duration
of function calls, memory usage, CPU usage, and other system resources along
with associated metadata.

## Benefits of Continuous Profiling

> Traditional profiling solutions have significant drawbacks limiting adoption in

Reviewer: What is "Traditional profiling", a reader new to this topic _might_ not know the difference between the one and the other. Is there an easy way to explain in a few words what this is and by that make it clearer where the drawbacks are coming from?

Reviewer 2: I'd say first define "Continuous Profiling" properly in the previous section.
Then "Traditional Profiling" is any other solution (non-continuous profiling).
So maybe just replace `Traditional` with `Other`?

Reviewer 2: And then I think the explanation of the differences is clarified in the following text.

Reviewer: I think "Traditional" is better than "Other", or even "Non-continuous profiling" would be fine, but ultimately we need to make clear what we mean by that, because from that the drawbacks become clear, otherwise we will have a hard time to argue that every "other" profiling is really not able to attach to a service at runtime, or is every "other" profiling solution really not able to get visibility into third-party libraries?

Author: > Traditional profiling solutions, typically used for one-off, development time
optimizations, have significant drawbacks limiting adoption in production
environments

Author: @svrnm what about that proposal 👆 ?

Reviewer: Is there something we can group them by like

```
Traditional profiling solutions rely on instrumentation code that needs to be inserted during compilation. They have significant drawbacks ...
```

Reviewer 2: "inserted during compilation" isn't true for all non-continuous profilers (e.g. profilers for scripted languages can be switched on/off via configuration or command-line flags). Also, you could have a continuous profiler that requires instrumentation.

Maybe we can summarize "traditional" as "(normally) require restarting of targeted apps or interpreters" or as "require changes of targeted apps" or "will interrupt production flow". You get the point, you'll surely find a better wording.

production environments:

- Significant cost and performance overhead due to code instrumentation
- Disruptive service restarts
- Inability to get visibility into third-party libraries

Continuous profiling, however, runs in the background with minimal overhead,
providing real-time, actionable insights without the need to replicate issues in
separate environments.

This allows SREs, DevOps, and developers to see how code affects performance and
cost, making code and infrastructure improvements easier.

## Contribution of comprehensive profiling abilities

> Elastic’s continuous profiling agent, [based on eBPF](https://ebpf.io/), is a

Reviewer: ```suggestion
The continuous profiling agent, that Elastic is donating, is [based on eBPF](https://ebpf.io/) and by that a
```

whole system, always-on solution that observes code and third-party libraries,
kernel operations, and other code you don't own. It eliminates the need for code
instrumentation (run-time/bytecode), recompilation, or service restarts with low
overhead, low CPU (~1%), and memory usage in production environments.

> The Elastic profiling agent facilitates identifying non-optimal code paths,

Reviewer: ```suggestion
The profiling agent facilitates identifying non-optimal code paths,
```

Similar to my comment above, since it's now donated to the OpenTelemetry community, should we drop the "Elastic" in front of it?

Author: how about "the donated profiling agent" ?

Reviewer: Yes, that sounds good to me.

uncovering "unknown unknowns", and provides comprehensive visibility into the

> runtime behavior of all applications. Elastic’s continuous profiling agent
> provides support for a wide range of runtimes and languages, such as:

Reviewer: ```suggestion
runtime behavior of all applications. The continuous profiling agent
provides support for a wide range of runtimes and languages, such as:
```

Apply this if the change above is accepted

- Minimize environmental impact: energy consumption associated with computing is
  a growing concern (source:
  [MIT Energy Initiative](https://energy.mit.edu/news/energy-efficient-computing/)).
  More efficient code translates to lower energy consumption, contributing to a
  reduction in carbon (CO2) footprint. Combining profiling data with
  OpenTelemetry's resource information (i.e. resource attributes) allows to
  derive insights into the services' carbon footprint.

- Through a detailed breakdown of services' resource utilization, profiling data
  provides actionable information on performance optimization opportunities.

- Improved vendor neutrality: a vendor-agnostic eBPF-based profiling agent
  removes the need to rely on proprietary agents to collect profiling telemetry.

With these benefits, SREs, developers, and DevOps, can now manage the overall
application’s efficiency on the cloud while ensuring their engineering teams
optimize it.

As the next step, Elastic and the OpenTelemetry profiling SIG will jointly work
on integrating the donated agent into OpenTelemetry's components ecosystem. We
look forward to providing a fully integrated and usable version of the new

> OpenTelemetry eBPF profiling agent to the users, soon.

Reviewer: Great closing statement, one _nit_:

"Elastic and the OpenTelemetry profiling SIG" makes it appear that Elastic stands outside of that SIG. Maybe we can rephrase that a little bit?

Continuous profiling, however, runs in the background with minimal overhead,
providing real-time, actionable insights without the need to replicate issues in
separate environments.

This allows SREs, DevOps, and developers to see how code affects performance and
cost, making code and infrastructure improvements easier.

## Contribution of comprehensive profiling abilities

The continuous profiling agent, that Elastic is donating, is
[based on eBPF](https://ebpf.io/) and by that a whole system, always-on solution
that observes code and third-party libraries, kernel operations, and other code
you don't own. It eliminates the need for code instrumentation
(run-time/bytecode), recompilation, or service restarts with low overhead, low
CPU (~1%), and memory usage in production environments.

The donated profiling agent facilitates identifying non-optimal code paths,
uncovering "unknown unknowns", and provides comprehensive visibility into the
runtime behavior of all applications. The continuous profiling agent provides
support for a wide range of runtimes and languages, such as:

- C/C++
- Rust
- Zig
- Go
- Java
- Python
- Ruby
- PHP
- Node.js / V8
- Perl
- .NET

> Additionally, organizations can combine profiling data with resource information
> to derive an estimate for the carbon footprints of services. This helps
> organizations meet their sustainability objectives by minimizing computational
> wastage, ensuring seamless alignment with their strategic ESG goals.

Reviewer: ```suggestion
```

This feels redundant based on the fact that this is listed as a benefit below. 

As I wrote in one of my previous comments on that topic: I think we need to be wary with hype topics like sustainability, artifical intelligence, etc and not over-sell of them. I would rather play it low in this blog post and have a blog post in a year from now that gives a deep dive into how you can use profiling to do exactly that.

## Benefits to OpenTelemetry

This contribution not only boosts the standardization of continuous profiling
for observability but also accelerates its adoption as a key signal in
OpenTelemetry. Customers benefit from a vendor-agnostic method of collecting
profiling data correlating it with existing signals, like tracing, metrics, and
logs, opening new potential for observability insights and a more efficient
troubleshooting experience.

### User Benefits of OpenTelemetry Profiling

OpenTelemetry-based continuous profiling unlocks the following possibilities for
users:

- Continuous profiling data compliments the existing signals (traces, metrics
  and logs) by providing detailed, code-level insights on the services'
  behavior.

- Seamless correlation with other OpenTelemetry signals such as traces,
  increasing fidelity and investigatory depth.

> - Minimize environmental impact: energy consumption associated with computing is
>   a growing concern (source:
>   [MIT Energy Initiative](https://energy.mit.edu/news/energy-efficient-computing/)).
>   More efficient code translates to lower energy consumption, contributing to a
>   reduction in carbon (CO2) footprint. Combining profiling data with
>   OpenTelemetry's resource information (i.e. resource attributes) allows to
>   derive insights into the services' carbon footprint.

Reviewer: ```suggestion
- Estimate environmental impact: Combining profiling data with
  OpenTelemetry's resource information (i.e. resource attributes) allows to
  derive insights into the services' carbon footprint.
```

Observability provides the data that allows to estimate, analyze, etc, but it does not minimize the environmental impact directly. "Enable minimization of environmental impact" would also be possible.

---

> title: Elastic Contributes its Continuous Profiling Agent to OpenTelemetry

Reviewer: ```suggestion
title: Elastic Contributes it's Continuous Profiling Agent to OpenTelemetry
```

Not a native speaker, so I might be wrong. Or drop the "it(')s" altogether, the title _should_ still say the same thing

Author: `it's` would mean `it is` ([see here](https://www.merriam-webster.com/grammar/when-to-use-its-vs-its#:~:text=It's%20is%20a%20contraction%20and,%2C%22%20for%20nouns%20without%20gender.)), so `its` should be fine here

Reviewer: right... 🙃

linkTitle: Elastic Contributes Profiling Agent # Mandatory, make sure that your short title.
date: 2024-06-06 # Put the current date, we will keep the date updated until your PR is merged
# prettier-ignore
cSpell:ignore: Bahubali Shetti Dmitry Filimonov Geisendörfer Halliday Christos Kalkanis
author:
  >- # If you have only one author, then add the single name on this line in quotes.
  [Bahubali Shetti](https://github.com/bshetti) (Elastic), [Alexander
  Wert](https://github.com/AlexanderWert) (Elastic), [Morgan
  McLean](https://github.com/mtwo) (Splunk), [Ryan
  Perry](https://github.com/Rperry2174) (Grafana)

> draft: true # TODO: remove this line once your post is ready to be published
> # canonical_url: http://somewhere.else/ # TODO: if this blog post has been posted somewhere else already, uncomment & provide the canonical URL here.

Reviewer: ```suggestion
```

This marks a significant milestone in establishing profiling as a core telemetry
signal in OpenTelemetry. Elastic’s [eBPF](https://ebpf.io/) profiling agent
observes code across different programming languages and runtimes, third-party
libraries, kernel operations, and system resources with low CPU and memory
overhead in production. Both, SREs and developers can now benefit from these
capabilities: quickly identifying performance bottlenecks, maximizing resource
utilization, reducing carbon footprint, and optimizing cloud spend.

Elastic’s decision to contribute the project to OpenTelemetry was made to
accelerate OpenTelemetry’s mission and enable effective observability through
high-quality, portable telemetry. This collaboration also shows the commitment
to vendor neutrality and community-driven development enhancing the overall
profiling and observability ecosystems.

The Elastic team would like to thank the OpenTelemetry community for the great
and constructive cooperation on the donation proposal. We look forward to
jointly establishing continuous profiling as an integral part of OpenTelemetry.

> 

Reviewer: Since this is the otel blog, shouldn't it be the other way round, otel thanking elastic? Otherwise the blog post is written by elastic contributors.

Author: rephrased

With today’s acceptance, Elastic’s continuous profiling agent will be
contributed to OpenTelemetry. This agent will now be jointly supported by both
Elastic’s team as well as a diverse set of official maintainers from different
companies:

- Dmitry Filimonov (Grafana Labs)
- Felix Geisendörfer (Datadog)
- Jonathan Halliday (Red Hat)
- Christos Kalkanis (Elastic)

## What is continuous profiling?

[Continuous profiling](https://www.cncf.io/blog/2022/05/31/what-is-continuous-profiling/)
is a technique used to understand the behavior of a software application by

> collecting information about its execution. This includes tracking the duration

Reviewer 2: This paragraph is also true for just `profiling` instead of `continuous profiling`.
```suggestion
collecting information about its execution over time. This includes tracking the duration
```
Alternative suggestion: `over time` -> `as a function of time`.
(Just ideas, not sure how a native speaker would phrase this, tbh.)

---
title: Elastic Contributes its Continuous Profiling Agent to OpenTelemetry
linkTitle: Elastic Contributes Profiling Agent # Mandatory, make sure that your short title.
date: 2024-06-06 # Put the current date, we will keep the date updated until your PR is merged
author:
  >- # If you have only one author, then add the single name on this line in quotes.
  [Bahubali Shetti](https://github.com/bshetti) (Elastic), [Alexander
  Wert](https://github.com/AlexanderWert) (Elastic), [Morgan
  McLean](https://github.com/mtwo) (Splunk), [Ryan
  Perry](https://github.com/Rperry2174) (Grafana)
draft: true # TODO: remove this line once your post is ready to be published
# canonical_url: http://somewhere.else/ # TODO: if this blog post has been posted somewhere else already, uncomment & provide the canonical URL here.

> issue: the issue ID for this blog post # TODO: See https://opentelemetry.io/docs/contributing/blog/ for details

Reviewer: Point to the donation issue here (community/<id>)

Reviewer 3: Do you want the issue ID or a link?

[Elastic profiling agent donation](https://github.com/open-telemetry/community/issues/1918)

what is the right semantic here?

sig: Profiling SIG
---

Following significant collaboration between
[Elastic](https://www.elastic.co/observability-labs/blog/elastic-donation-proposal-to-contribute-profiling-agent-to-opentelemetry)
and
[OpenTelemetry's profiling community](https://opentelemetry.io/blog/2024/profiling/),
which included a thorough review process, we’re excited to announce that the

> OpenTelemetry project has accepted Elastic's donation of its continuous

Reviewer: Add a link to the donation issue

profiling agent.

This marks a significant milestone in establishing profiling as a core telemetry

> signal in OpenTelemetry. Elastic’s eBPF profiling agent observes code across

Reviewer: Link to the ebpf project website when ebpf is mentioned the first time

- Inability to get visibility into third-party libraries

Continuous profiling, however, runs in the background with minimal overhead,
providing real-time, actionable insights without the need to replicate issues in
separate environments.

This allows SREs, DevOps, and developers to see how code affects performance and
cost, making code and infrastructure improvements easier.

## Contribution of comprehensive profiling abilities

Elastic’s continuous profiling agent, [based on eBPF](https://ebpf.io/), is a
whole system, always-on solution that observes code and third-party libraries,
kernel operations, and other code you don't own. It eliminates the need for code
instrumentation (run-time/bytecode), recompilation, or service restarts with low
overhead, low CPU (~1%), and memory usage in production environments.

The Elastic profiling agent facilitates identifying non-optimal code paths,
uncovering "unknown unknowns", and provides comprehensive visibility into the
runtime behavior of all applications. Elastic’s continuous profiling agent
provides support for a wide range of runtimes and languages, such as:

- C/C++
- Rust
- Zig
- Go
- Java
- Python
- Ruby
- PHP
- Node.js

> - V8

Reviewer: Isn't nodejs built on top of v8. Is there a reason to mention both?

Reviewer 3: Will remove V8

- Perl
- .NET

Additionally, organizations can meet sustainability objectives by minimizing
computational wastage, ensuring seamless alignment with their strategic ESG

> goals.

Reviewer: Is there any tangible proof that profiling is doing that already otherwise I would remove the wording around sustainability, it's a buzzword right now and we should avoid sounding "trendy" without evidence (let's have a dedicated blog post on how to do this later on)

Author: @svrnm Based on the data the agent we're doniting provides, we provide insights on the CO2 usage already in the Elastic backend. So, this is tangible and is something that can already be derived from that data.

Author: @svrnm we changed the wording to make it a bit more concrete.

  profitability by reducing the computational resources needed to run
  applications. Whole system continuous profiling identifies the most expensive
  functions (down to the lines of code) across diverse environments that may
  span multiple cloud providers. In the cloud context, every CPU cycle saved
  translates to money saved.

- Minimize environmental impact: energy consumption associated with computing is
  a growing concern (source:
  [MIT Energy Initiative](https://energy.mit.edu/news/energy-efficient-computing/)).
  More efficient code translates to lower energy consumption, contributing to a
  reduction in carbon (CO2) footprint.

- Accelerate engineering workflows: continuous profiling provides detailed
  insights to help troubleshoot complex issues faster, guide development, and
  improve overall code quality.

- Improved vendor neutrality: a vendor-agnostic eBPF-based profiling agent
  removes the need to use proprietary APM agents to collect profiling telemetry.

With these benefits, SREs, devs, and DevOps, can now manage the overall
application’s efficiency on the cloud while ensuring their engineering teams

> optimize it.

Reviewer: This speaks to a buyer persona / management, but our audience are technical, so I would rather prefer a different set of advantages. I think it would be especially important to have make clear how profiling complements the other signals

Reviewer 3: Sure, I would suggest we use these:

- Improved customer experience: delivering consistent service quality and
  performance through continuous profiling ensures customers have an application
  that performs optimally, remains responsive, and is reliable.

- Accelerate engineering workflows: continuous profiling provides detailed
  insights to help troubleshoot complex issues faster, guide development, and
  improve overall code quality.

- Improved vendor neutrality: a vendor-agnostic eBPF-based profiling agent
  removes the need to use proprietary APM agents to collect profiling telemetry.

As an SRE these are the simplest advantages that make sense out of that list

Reviewer 3: Let me know what you think

Author: @svrnm revised this section with a set of more technical benefits

Reviewer: Thanks!

Elastic remains deeply committed to the
[OpenTelemetry vision and mission](https://opentelemetry.io/community/mission/)
and will continue to contribute and participate in profiling and other areas of

> OpenTelemetry.

Reviewer: We should close with some "joined statement" or an otel focused wording, since this is the otel not the elastic blog

Reviewer 3: What do you suggest?

Author: added a proposal for a joint statement
